rust/lance-file/src/v2/reader.rs

```diff
@@ -218,23 +218,34 @@ impl ReaderProjection {
     ///
     /// If the schema provided is not the schema of the entire file then
     /// the projection will be invalid and the read will fail.
+    /// If the field is a `struct datatype` with `packed` set to true in the field metadata,
+    /// the whole struct has one column index.
+    /// To support nested `packed-struct encoding`, this method need to be further adjusted.
     pub fn from_whole_schema(schema: &Schema, version: LanceFileVersion) -> Self {
         let schema = Arc::new(schema.clone());
         let is_structural = version >= LanceFileVersion::V2_1;
-        let mut counter = 0;
-        let counter = &mut counter;
-        let column_indices = schema
-            .fields_pre_order()
-            .filter_map(|field| {
-                if field.children.is_empty() || !is_structural {
-                    let col_idx = *counter;
-                    *counter += 1;
-                    Some(col_idx)
-                } else {
-                    None
-                }
-            })
-            .collect::<Vec<_>>();
+        let mut column_indices = vec![];
+        let mut curr_column_idx = 0;
+        let mut packed_struct_fields_num = 0;
+        for field in schema.fields_pre_order() {
+            let field_metadata = &field.metadata;
+            if packed_struct_fields_num > 0 {
+                packed_struct_fields_num -= 1;
+                continue;
+            }
+            if field_metadata
+                .get("packed")
+                .map(|v| v == "true")
```

Comment:
Let's make this a case insensitive comparison

Reply:
fixed.